# Assignment 1
## Medical Patient No Show

### Team members: Luay Dajani, Dana Geislinger, Chris Morgan, Caroll Rodriguez
##### Github - https://github.com/cdmorgan103/7331DataMiningNoShow

## Business Understanding
We utilized the Kaggle No Show appointment data that contains the show/no-show status for several clinics across Brazil that are a part of a central healthcare system. The data set has 110,527 records and 14 attributes. This data was collected in 2016 and shows the show/no-show behavior of the patients across the different medical facilities (that are identified as the Neighborhood location).  

No Show in the medical world can be a large cost for clinics and health systems. With healthcare institution margins shrinking and an effort to modernize the delivery of healthcare, prioritizing nonproductive time has become a more common focus. As the institution's capacity becomes more constrained and scheduling time frames grow into the weeks, no show is often cited as becoming an increasingly important issue and many medical systems have gone to extreme lengths to reduce these costly occurrences. 

This has led many institutions to take various actions such as appointment confirmation via SMS, consumer engagement via email/web, strict planned overbooking rates, and even predictive modeling (my healthcare employer has done this).

In the case of predictive modeling, we have to not only think of how much yield our model can reasonably contribute, but we also have to consider the impact of a predicted No-show. Are we going to overbook for predicted no-show patients? If we predict No-show too often and have "Show patients", there is the potential for serious ramifications in the form of clinical overload. Therefore, it is critical to ensure that we consider how the model will be utilized and consider being as conservative as possible for no-show prediction.

As data scientists, it will be critical for us to not only explain model performance, but coach business leaders on how the model will work and be impacted by changes. This means establishing an effective communication channel between business leaders and the Data Science team so impacts of any operational changes (i.e. changing scheduling practices, adding pre-appointment verification calls, etc.) can be communicated. Furthermore, this may mean helping to coach business leaders on how they can use the model and where to put resources to improve business performance (i.e. following up with high probability no-show appointments before their appointment).




## Data Meaning Type

### Data Meaning
This dataset contains 110,527 appointment records for clinics located across the coastal city of Vitória in Espírito Santo, Brazil. The dataset includes 11 meaningful predictors relating to each appointment and to the patient that scheduled that appointment. Unique numeric identifiers are provided for each patient as well as for each appointment. The response variable of interest for this data set, *No-show*, is a boolean variable denoting whether or not a patient made it to their scheduled appointment.

Of the 11 predictor variables (the 2 unique identifiers are excluded), there are 2 timepoints, 1 integer, 1 categorical, and 7 boolean variables. Data are provided that generally describe a patient's health problems as well as their age, gender, and scholastic background. In addition to patient information, we are given data pertaining to the location and time of the appointment as well as whether or not the patient was notified about their appointment with an automated SMS reminder.

The following table describes all the variables provided in greater detail. Perhaps the most intriguing variable is *Scholarship*, which is related to the Bolsa Família social program instituted in Brazil. According to the dataset curator, "the best explanation about this variable that indicates if the person receives a scholarship or not" (source: https://www.kaggle.com/joniarroba/noshowappointments/discussion/45899). Based on this description, we assume that this variable defines whether or not a patient is currently receiving financial aid as part of this social program. Participants in Bolsa Família must have an income of less than $170/month, must attend regular medical checkups for all mothers and children in the household, and children must regularly attend school (source: https://www.wilsoncenter.org/article/programa-bolsa-familia).

| Variable Name  | Data Type | Variable Type         | Description                                                             |
| -------------- | --------- | --------------------- | ----------------------------------------------------------------------- |
| PatientID      | Interval  | Identifier            | Unique ID number for each patient.                                      |
| AppointmentID  | Interval  | Identifier            | Unique ID number for each appointment.                                  |
| Gender         | Nominal   | Binary Predictor      | Sex of the patient (Male/Female).                                       |
| ScheduledDay   | Interval  | Date/Time Predictor   | **Date** and **Time** when the patient called to schedule their appointment. Should always be before *AppointmentDay*.                                                                         |
| AppointmentDay | Interval  | Date Predictor        | Scheduled appointment **Date**. Appointment **Times** are not provided. |
| Age            | Ratio     | Integer Predictor     | Age of the patient in years.                                            |
| Neighbourhood  | Nominal   | Categorical Predictor | The neighborhood in which the appointment facility is located.          |
| Scholarship    | Ordinal   | Boolean Predictor     | Whether or not the patient receives Bolsa Família financial aid. To receive this benefit, a patient's income must be under the poverty threshold, all children in the household must be vaccinated and regularly attending school, and mothers and children must receive routine medical care.                                    |
| Hipertension   | Ordinal   | Boolean Predictor     | Whether or not a patient is classified as hypertensive (has high blood pressure).                                                                                                                     |
| Diabetes       | Ordinal   | Boolean Predictor     | Whether or not a patient is diagnosed as a diabetic.                    |
| Alcoholism     | Ordinal   | Boolean Predictor     | Whether or not a patient is classified as an alcoholic.                 |
| Handcap        | Ordinal   | Boolean Predictor     | Whether or not a patient is diagnosed as being handicapped.             |
| SMS_received   | Nominal   | Boolean Predictor     | Whether or not a patient received an SMS (text message) reminder for   their appointment.                                                                                                             |
| No-show        | Nominal   | Boolean Response      | Whether or not a patient showed up for their appointment. True means they **did not** show up, False means they **did** show up.                                                                         |

#### Created Variables
| Variable Name  | Data Type | Variable Type         | Description                                                             |
| -------------- | --------- | --------------------- | ----------------------------------------------------------------------- |
| DaysInAdvance  | Ratio     | Integer Predictor     | Value for how many days in advance the appointment was scheduled.       |
| ScheduledDOW   | Nominal   | Categorical Predictor | Day of the week for the day the patient scheduled the appointment.      |
| AppointmentDOW | Nominal   | Categorical Predictor | Day of the week for patient appointment.                                |
| ScheduledTime  | Interval  | Time Predictor        | **Time** of day when an appointment was scheduled.                      |

Dataset from: https://www.kaggle.com/joniarroba/noshowappointments

### Verify Data Quality

To verify the data quality, we will first import the raw data as a Pandas DataFrame object. Next, we will determine whether or not the data requires cleaning or modifications before deeper analysis.  This process will include changing variable types and names to more practically useful formats. Finally, we will perform basic exploratory data analysis to elucidate any potential patterns or trends within the data.

In [1]:
# Import required modules
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib import style
from pprint import pprint
from IPython.display import display

# Load the data into variable 'df'
df = pd.read_csv('data/KaggleV2-May-2016.csv')

# Get an overview of the raw data
df.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
PatientId         110527 non-null float64
AppointmentID     110527 non-null int64
Gender            110527 non-null object
ScheduledDay      110527 non-null object
AppointmentDay    110527 non-null object
Age               110527 non-null int64
Neighbourhood     110527 non-null object
Scholarship       110527 non-null int64
Hipertension      110527 non-null int64
Diabetes          110527 non-null int64
Alcoholism        110527 non-null int64
Handcap           110527 non-null int64
SMS_received      110527 non-null int64
No-show           110527 non-null object
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


With the raw data imported, we see that there are already no null (missing) values for any of the 110,527 observations in the data set. However, most of the categorical or binary variables are incorrectly stored as generic numpy objects or 64-bit integers. Furthermore, the date-time columns are stored as generic numpy objects. These columns should be converted to the correct data types. Finally, *ScheduledDay* includes time of day, but *AppointmentDay* does not, so we will separate the Scheduled time of day into a new variable; this way, it will be easy to compare the scheduled/actual **days**, while still retaining the time of day at which each appointment scheduling occurred.

In [2]:
# Convert categorical variables to the correct datatype
categ_features = ['Gender', 'Neighbourhood', 'Scholarship', 'Hipertension', 'Diabetes',
                  'Alcoholism', 'Handcap', 'SMS_received', 'No-show'
                  ]
df[categ_features] = df[categ_features].astype('category')

# Pull-out the scheduled time of day as a new variable (ScheduledTime) and re-insert into df
df.ScheduledDay, ScheduledTime = df.ScheduledDay.str.split('T', 1).str
df.insert(loc=4, column='ScheduledTime', value=ScheduledTime)

# Convert date-time variables to the correct type using the C-style fmt codes
df.ScheduledDay = pd.to_datetime(df.ScheduledDay, format="%Y-%m-%d")
df.ScheduledTime = pd.to_datetime(df.ScheduledTime, format="%H:%M:%SZ").dt.time
df.AppointmentDay = pd.to_datetime(df.AppointmentDay, format="%Y-%m-%dT%H:%M:%SZ")

# Reprint df info to check
df.head()

PatientId  AppointmentID Gender ScheduledDay ScheduledTime  \
0  2.987250e+13        5642903      F   2016-04-29      18:38:08   
1  5.589978e+14        5642503      M   2016-04-29      16:08:27   
2  4.262962e+12        5642549      F   2016-04-29      16:19:04   
3  8.679512e+11        5642828      F   2016-04-29      17:29:31   
4  8.841186e+12        5642494      F   2016-04-29      16:07:23   

  AppointmentDay  Age      Neighbourhood Scholarship Hipertension Diabetes  \
0     2016-04-29   62    JARDIM DA PENHA           0            1        0   
1     2016-04-29   56    JARDIM DA PENHA           0            0        0   
2     2016-04-29   62      MATA DA PRAIA           0            0        0   
3     2016-04-29    8  PONTAL DE CAMBURI           0            0        0   
4     2016-04-29   56    JARDIM DA PENHA           0            1        1   

  Alcoholism Handcap SMS_received No-show  
0          0       0            0      No  
1          0       0            0      No  
2          0       0            0      No  
3          0       0            0      No  
4          0       0            0      No

The *PatientId* column was cast as a floating point variable, because there are some rows that erroneously contain data after the decimal point. We will convert these rows to integers and check if the ID numbers are unique for each row. It should be noted that it would not be incorrect to have multiple identical patient ID's, as these would describe multiple appointments made for the same patient. However, appointment ID should be unique for each observation. We will check these values before and after the conversion of *PatientId* to make sure that each patient is correctly identified after the conversion.

In [3]:
# Print the number of unique appointment IDs
print("Unique Appointment IDs: %d" %  len(df.AppointmentID.unique()))
print("Total Appointments:     %d\n" % len(df.AppointmentID))

# Print the number of unique patient IDs
print("Unique Patient IDs Before Conversion: %d" % len(df.PatientId.unique()))

# Cast PatientId as int
df.PatientId = df.PatientId.astype(np.int64)

# Double check the counts
print("Unique Patient IDs After Conversion:  %d" % len(df.PatientId.unique()))

Unique Appointment IDs: 110527
Total Appointments:     110527

Unique Patient IDs Before Conversion: 62299
Unique Patient IDs After Conversion:  62299


This tells us that each appointment is indeed uniquely identified, but that all appointments in the dataset were made by the same group of 62,299 patients.

#### Cleaning the data
This data set is relatively clean in its raw state, with no missing values in any columns. However, there are some inconsistencies and errors that must be corrected before further analysis can be performed:

Since this dataset originated in a non-English speaking country, there are some columns with names that are misspelled. There are also some columns that are named inconsistently from the rest of the data. These names will be changed before further analysis is performed.

In [4]:
# Rename incorrect column names.
df = df.rename(columns={'Hipertension': 'Hypertension', 'Handcap': 'Handicap', 'SMS_received': 'SMSReceived', 'No-show': 'NoShow'})
pprint(list(df.columns))

['PatientId',
 'AppointmentID',
 'Gender',
 'ScheduledDay',
 'ScheduledTime',
 'AppointmentDay',
 'Age',
 'Neighbourhood',
 'Scholarship',
 'Hypertension',
 'Diabetes',
 'Alcoholism',
 'Handicap',
 'SMSReceived',
 'NoShow']


There are sub-zero values for age and two unique patients with an age of 115. This could be documented oddly for perhaps a pregnant mother's child (hence a negative age value since the child isn't born). Since we have no way of knowing if why this single data point was recorded as -1, the -1 value will be simply imputed with the median value for the dataset. It is unlikely the 2 patients who are actually 115, but it is feasible to reach this age, and most likely, these patients true age are within the vicinity of the age documented but are slightly off due to poor birth documentation. In this case the values of 115 and other suspiciously high age values will remain the same at this point.  

In [5]:
# Print observations where age is minimum or maximum
display(df.loc[(df.Age == -1) | (df.Age == 115)])

# Impute the values of sub zero age observations
df.Age=df.Age.replace(-1, int(df.Age.median()))

PatientId  AppointmentID Gender ScheduledDay ScheduledTime  \
63912   31963211613981        5700278      F   2016-05-16      09:17:44   
63915   31963211613981        5700279      F   2016-05-16      09:17:44   
68127   31963211613981        5562812      F   2016-04-08      14:29:17   
76284   31963211613981        5744037      F   2016-05-30      09:44:51   
97666  748234579244724        5717451      F   2016-05-19      07:57:56   
99832  465943158731293        5775010      F   2016-06-06      08:58:13   

      AppointmentDay  Age Neighbourhood Scholarship Hypertension Diabetes  \
63912     2016-05-19  115    ANDORINHAS           0            0        0   
63915     2016-05-19  115    ANDORINHAS           0            0        0   
68127     2016-05-16  115    ANDORINHAS           0            0        0   
76284     2016-05-30  115    ANDORINHAS           0            0        0   
97666     2016-06-03  115      SÃO JOSÉ           0            1        0   
99832     2016-06-06   -1         ROMÃO           0            0        0   

      Alcoholism Handicap SMSReceived NoShow  
63912          0        1           0    Yes  
63915          0        1           0    Yes  
68127          0        1           0    Yes  
76284          0        1           0     No  
97666          0        0           1     No  
99832          0        0           0     No

#### Defining New Variables
Next, we determine how many days in advance the appointment was scheduled and create it as a new feature. This will ensure values are reasonable and that there are no negative data points. Examination shows we do have some data points that need to be corrected, since scheduling an appointment for a previous date should not be possible.

In [6]:
# Create a column showing days in advance
df['DaysInAdvance']=(df['AppointmentDay']-df['ScheduledDay']).dt.days

# List appointments with negative days in advance (logically impossible)
df.loc[df.DaysInAdvance<0]

PatientId  AppointmentID Gender ScheduledDay ScheduledTime  \
27033    7839272661752        5679978      M   2016-05-10      10:51:53   
55226    7896293967868        5715660      F   2016-05-18      14:50:41   
64175   24252258389979        5664962      F   2016-05-05      13:43:58   
71533  998231581612122        5686628      F   2016-05-11      13:49:20   
72362    3787481966821        5655637      M   2016-05-04      06:50:57   

      AppointmentDay  Age  Neighbourhood Scholarship Hypertension Diabetes  \
27033     2016-05-09   38    RESISTÊNCIA           0            0        0   
55226     2016-05-17   19  SANTO ANTÔNIO           0            0        0   
64175     2016-05-04   22     CONSOLAÇÃO           0            0        0   
71533     2016-05-05   81  SANTO ANTÔNIO           0            0        0   
72362     2016-05-03    7     TABUAZEIRO           0            0        0   

      Alcoholism Handicap SMSReceived NoShow  DaysInAdvance  
27033          0        1           0    Yes             -1  
55226          0        1           0    Yes             -1  
64175          0        0           0    Yes             -1  
71533          0        0           0    Yes             -6  
72362          0        0           0    Yes             -1

Since scheduling an appointment on a later day than the appointment is scheduled for is impossible, we will impute the *ScheduledDay* for all appointments with *DaysInAdvance* < 0 as the same day as the appointment. This seems to be a logical imputation since many of the existing appointments in the data set were scheduled the same day as the appointment, and appointments with scheduled days in the future are likely data entry errors.

In [7]:
# Run through the data to ensure no appointments that are scheduled after the appointment(which would be impossible).
# If true, scheduled day with the appointment day is assumed as the a same day as the appointment, then recalculate advance field
df['ScheduledDay'] = np.where(df['ScheduledDay']>df['AppointmentDay'], df['AppointmentDay'], df['ScheduledDay'])
df['DaysInAdvance']=(df['AppointmentDay']-df['ScheduledDay']).dt.days

#Examine again, we have corrected the bad scheduled appointment data (table is now empty)
df.loc[df.DaysInAdvance<0]

Empty DataFrame
Columns: [PatientId, AppointmentID, Gender, ScheduledDay, ScheduledTime, AppointmentDay, Age, Neighbourhood, Scholarship, Hypertension, Diabetes, Alcoholism, Handicap, SMSReceived, NoShow, DaysInAdvance]
Index: []

Next, we will add variables to the data set that describe the day of the week of an appointment and the day of the week on which an appointment was scheduled. The day of the week could provide important insights into trends in the data that might go unnoticed as raw dates.

In [8]:
# Create a day of week variable for both the scheduled day and the appointment day which will allows to examining
#  any potential trends related to the day of the week and appointment no-show
df['ScheduledDOW'] = df['ScheduledDay'].dt.weekday_name
df['AppointmentDOW'] = df['AppointmentDay'].dt.weekday_name

#Check the variables
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 18 columns):
PatientId         110527 non-null int64
AppointmentID     110527 non-null int64
Gender            110527 non-null category
ScheduledDay      110527 non-null datetime64[ns]
ScheduledTime     110527 non-null object
AppointmentDay    110527 non-null datetime64[ns]
Age               110527 non-null int64
Neighbourhood     110527 non-null category
Scholarship       110527 non-null category
Hypertension      110527 non-null category
Diabetes          110527 non-null category
Alcoholism        110527 non-null category
Handicap          110527 non-null category
SMSReceived       110527 non-null category
NoShow            110527 non-null category
DaysInAdvance     110527 non-null int64
ScheduledDOW      110527 non-null object
AppointmentDOW    110527 non-null object
dtypes: category(9), datetime64[ns](2), int64(4), object(3)
memory usage: 8.5+ MB


PatientId  AppointmentID            Age  DaysInAdvance
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000
mean   1.474963e+14   5.675305e+06      37.089218      10.183792
std    2.560949e+14   7.129575e+04      23.109921      15.254924
min    3.921700e+04   5.030230e+06       0.000000       0.000000
25%    4.172614e+12   5.640286e+06      18.000000       0.000000
50%    3.173184e+13   5.680573e+06      37.000000       4.000000
75%    9.439172e+13   5.725524e+06      55.000000      15.000000
max    9.999816e+14   5.790484e+06     115.000000     179.000000

With the data cleaned, we print out the unique values for each categorical variable and the ranges for continuous variables to identify patterns and to better understand our data. This also allows us to verify that the contents of the data are as we understand them.

In [9]:
# Print descriptive info for the unique values for each predictor
print('Gender:', list(df.Gender.unique()))
print('Scheduled Dates: %s to %s' % (min(df.ScheduledDay).date(), max(df.ScheduledDay).date()))
print('Appointment Dates: %s to %s' % (min(df.AppointmentDay).date(), max(df.AppointmentDay).date()))
print('Age Range: %d to %d Years Old' % (min(df.Age), max(df.Age)))
print('Number of Distinct Neighbourhoods:', len(df.Neighbourhood.unique()))
print('Scholarship:', list(df.Scholarship.unique()))
print('Hypertension:', list(df.Hypertension.unique()))
print('Diabetes:', list(df.Diabetes.unique()))
print('Alchoholism:', list(df.Alcoholism.unique()))
print('Handicap:', list(df.Handicap.unique()))
print('SMSReceived:', list(df.SMSReceived.unique()))
print('NoShow:', list(df.NoShow.unique()))
print('Range of Scheduled Days in Advance: %d to %d Days' % (min(df.DaysInAdvance), max(df.DaysInAdvance)))
print('ScheduledDOW:', sorted(df.ScheduledDOW.unique()))
print('AppointmentDOW:', sorted(df.AppointmentDOW.unique()))

Gender: ['F', 'M']
Scheduled Dates: 2015-11-10 to 2016-06-08
Appointment Dates: 2016-04-29 to 2016-06-08
Age Range: 0 to 115 Years Old
Number of Distinct Neighbourhoods: 81
Scholarship: [0, 1]
Hypertension: [1, 0]
Diabetes: [0, 1]
Alchoholism: [0, 1]
Handicap: [0, 1, 2, 3, 4]
SMSReceived: [0, 1]
NoShow: ['No', 'Yes']
Range of Scheduled Days in Advance: 0 to 179 Days
ScheduledDOW: ['Friday', 'Monday', 'Saturday', 'Thursday', 'Tuesday', 'Wednesday']
AppointmentDOW: ['Friday', 'Monday', 'Saturday', 'Thursday', 'Tuesday', 'Wednesday']


Viewing the ranges/categories for each variable helps illuminates the data and conveys some interesting trends.

First, the date ranges show that appointments were scheduled as early as November 2015, but this dataset does not include any actual appointments prior to the end of April 2016. The latest appointment takes place on June 8th, 2016 and this is also the final date on which an appointment was scheduled.

It is interesting to note the large age range of patients in the data set, and important to note that the patients in this data set could have visited clinics in any one of 81 possible Brazilian neighborhoods.

It is also interesting to note that patients schedule appointments over a wide range, from the same day to almost a half a year in advance. Finally, it is very interesting to notice that there were both no appointments scheduled or visits taken on Sundays throughout the entire dataset. This seems to indicate that medical clinics in Brazil are closed Sunday, both for scheduling appointments and seeing patients.

In [10]:
# Count the number of rows in which handicap takes on each value
display(df.groupby(df.Handicap).Handicap.count())

# Count the number of Handicap > 1 appointments
print('%d appointments have Handicap > 1' % len(df.loc[(df.Handicap == 2) | (df.Handicap == 3) | (df.Handicap == 4)]))

Handicap
0    108286
1      2042
2       183
3        13
4         3
Name: Handicap, dtype: int64

199 appointments have Handicap > 1


Looking at the variable ranges in this way shows us one final issue with the data that wasn't noticed previously; the *Handicap* variable is meant to be a boolean feature denoting whether a patient is disabled (1) or not (0). However, there are values of 2, 3, 4, and 5 entered in for this data as in 199 rows as well.

We believe that it is likely that clinics erroneously input a handicap value on an ordinal scale from 0-4 instead of as a boolean 0/1 value. Since these observations comprise only a small portion (199 / 110,527) of the data, it could make sense to impute them. However, since we are not yet at the stage of building a model, we will leave the values as they are for now. Even with such a small portion of the data represented, it *Handicap* levels > 1 might turn out to be beneficial predictors.

Now that the raw data has been cleaned and variables have been investigated, we will save the updated data set to a new csv file.

In [ ]:
df.to_csv("data/updated.csv", index=False)

### Simple Statistics and Visualizing Attributes
This part will perform the most simplistic steps while doing the Exploratory Data Analysis. When describing the data, as there are either continuous, categorical (including binary) data, the system will not show them together, so their summary will be displayed in seperately as below:

#### AppointmentID
This is a unique designation, and there is nothing really interesting to talk about this.
#### Patient Id
As per the summary of Patient Id, there does not seem to be any specific trend as below:

In [ ]:
df['PatientId'].describe().apply(lambda x: format(x, '.0f'))

But as per deeper analysis, the number of unique Patient Id's are less than the that of the total observations, meaning that that the patients are taking multiple appointments as shown below:

In [ ]:
len(df.PatientId.unique())

This may mean that a Patient's history of skipping appointments can be calculated and potentially used to help predict future behavior.  Below is the top rows for a new dataframe calculated to collect each patient, number of appointments, and the percent of truancy:

In [ ]:
#Create a temporary dataset to add a temp variable of NoShow as a number (0=False,1=True)
df_temp = pd.DataFrame()
df_temp['PatientId'] = df['PatientId']
df_temp['NoShow'] = df['NoShow']
df_temp['NoShow_num'] = np.where(df_temp['NoShow']=='Yes', 1, 0)

#Create a subset dataframe which is basically a table pivoted (groupby) on PatientID
df_patient = pd.DataFrame()
df_patient['Appointments'] = df_temp.groupby(['PatientId'])['NoShow_num'].count()
df_patient['NoShow_sum'] = df_temp.groupby(['PatientId'])['NoShow_num'].sum()
df_patient['NoShow_percent'] = df_temp.groupby(['PatientId'])['NoShow_num'].sum() / df_temp.groupby(['PatientId'])['NoShow_num'].count()
df_patient['PatientID'] = df_temp.groupby(['PatientId'])['NoShow_num'].count().index
#reorder to put the PatientID first for aestetics b/c it would not work if I did this operation first
df_patient= df_patient.loc[0:,('PatientID','Appointments','NoShow_sum','NoShow_percent')]

df_patient[100:110]

#### Age and DaysInAdvance

In [ ]:
df.loc[0:,('Age', 'DaysInAdvance')].describe()

Referencing the table above, as discussed previously the minimum age starts 0 years of age -- assuming a new born -- and the maximum ends up being 115 years of age with the average age of approximately 37.  The interesting number is the standard deviation which is at only 23, likely indicating that there is a heavy concentration of patient ages around the mean and the high values are basically the outliers.  The same, but in a more extreme scenario, can be said about the concentration of people making appointments in advanced, where the mean is 10 days, but with an extreme instance of someone making an appointment 179 days in advanced.  These trends are indicated by the 1st and 3rd quartiles, but demonstrated with the box and violin plots below, the distributions are fat bottomed... meaning that for *Age*, most of the patients are younger; and for *DaysInAdvance*, most people tend not to schedule their appointments too much in advance, but their are a very few exceptions of people who do.

In [ ]:
import matplotlib.pyplot as plt
import warnings
warnings.simplefilter('ignore', DeprecationWarning)

#plot regular boxplot below others
ax = df.loc[0:,('Age', 'DaysInAdvance')].boxplot()
plt.figure() #Forces the boxplot to plot before the next ones
import seaborn as sns
#Create a side by side violin plot
for fi,feature in enumerate(['Age','DaysInAdvance']):
    plt.subplot(2,2,fi+1) #The 2,2 is the biggest size
    sns.violinplot(y=feature,data=df, split=True, inner='quart')

As per the rest of the variables, their summary statistics as below do not really reveal any intersting statistics, likley better to examine them separately.

In [ ]:
df.loc[0:,('Gender', 'ScheduledDay', 'ScheduledTime', 'AppointmentDay', 
 'Neighbourhood', 'Scholarship', 'Hypertension', 'Diabetes', 'Alcoholism', 'Handicap', 'SMSReceived',
 'NoShow')].describe()

#### NoShow
Probably the most important variable -- and more so as the main response variable -- is the frequency of NoShow.  As per the following breakdown from this sample, around 20.2% of total appointments are no shows.  On the surface this can be used as a general indicator, but more digging has to be done to find out which other variables contribute more or less to appointments being missed.

In [ ]:
show_values = [len(df.NoShow[df.NoShow=='Yes']), len(df.NoShow[df.NoShow!='Yes'])]
show_labels = ['No Show', 'Show']
plt.pie(show_values, labels = show_labels, autopct='%.1f')

#### Gender Breakdown
In this sample, as per the breakdown below, 65% from this sample are Female, and 35% are Male.

In [ ]:
#Breakup gender to female & male & plot pie chart
values = [len(df.Gender[df.Gender=='F']), len(df.Gender[df.Gender!='F'])]
gender_labels = ['Female', 'Male']
plt.pie(values, labels = gender_labels, autopct='%.1f')

#### Neighborhood
The breakdown of patients from the different barrios in Vitoria are as below with Jardin Camburi and Maria Ortiz as the top neighborhoods where most of the appointments were made.  Jardin Camburi is the most populous neighborhood (39,157 *Wikipedia: https://pt.wikipedia.org/wiki/Jardim_Camburi) in Vitoria (358,267 *wikipedia: https://pt.wikipedia.org/wiki/Vit%C3%B3ria_(Esp%C3%ADrito_Santo) ) or 10%, while, as per the numbers above taking up around the percentage as below (making a quick determination that the more populous neighborhoods likely represented more appointments)

In [ ]:
#Breakup Neighborhoods
df_neighborhood = df.groupby(['Neighbourhood'])['Neighbourhood'].count()
df_neighborhood = df_neighborhood.sort_values()
#Jardin Camburi percentage
total_appointments = sum(df_neighborhood.values)
JardinCamburi_appointments = df_neighborhood['JARDIM CAMBURI']
print("The total appointments in JARDIM CAMBURI was {} out of a total {} or {:.2f}%."
      .format(JardinCamburi_appointments,total_appointments,JardinCamburi_appointments/total_appointments*100))

Below is a full representation of the breakdown of appointments per neighborhood.

In [ ]:
#plot and format bar chart
plt.figure(figsize=(16,5), dpi=80)
df_neighborhood.plot(kind='bar')
#plt.barh(df_neighborhood, width=0.4, height=0.4, align='center', alpha=0.4)

### Explore Joint Attributes

In [ ]:
names = list(df.columns.values)
# plot correlation matrix
correlations = df.corr(method='pearson')
fig = plt.figure()
ax = fig.add_subplot(111)
cax = ax.matshow(correlations, vmin=-1, vmax=1)
fig.colorbar(cax)
ticks = np.arange(0,9,1)
ax.set_xticks(ticks)
ax.set_yticks(ticks)
ax.set_xticklabels(names, rotation=90)
ax.set_yticklabels(names)
plt.show()

In [ ]:
# Scatterplot Matrix - pairwise relationships
from pandas.plotting import scatter_matrix
scatter_matrix(df)
plt.show()

In [ ]:
import seaborn as sns
sns.set(style="dark")
#df = sns.load_dataset(df)
sns.pairplot(df[tuple([slice(None),1])], hue="NoShow")

##### Frequency Tables 

In [ ]:
df = df.replace({'NoShow': {'Yes': True, 'No': False}})

In [ ]:
#Ref: http://songhuiming.github.io/pages/2016/07/12/python-vs-sas/
#Gender
#Chris suggested I count unique PatientId
x=df.PatientId
x.value_counts(dropna = False).sort_index()
pd.crosstab(df.NoShow,df.Gender).apply(lambda r: r/r.sum(), axis = 1)

In [ ]:
#Frequency of missing an appointment based on Appointment Day of Week

pd.crosstab(df.NoShow,df.AppointmentDOW).apply(lambda r: r/r.sum(), axis = 1)

In [ ]:
#Frequency of missing an appointment based on the day of the week the appointment was made

pd.crosstab(df.NoShow,df.ScheduledDOW).apply(lambda r: r/r.sum(), axis = 1)

In [ ]:
# let's break up the age variable
df['age_range'] = pd.cut(df.Age,[0,16,65,1e6],3,labels=['child','adult','senior']) # this creates a new variable
df.age_range.describe()

In [ ]:
df.age_range

In [ ]:
#now lets group with the new variable
df_grouped = df.groupby(by=['age_range','NoShow'])
print ("Percentage of No Shows in each age group:")
print (df_grouped.NoShow.sum() / df_grouped.PatientId.nunique() *100)

In [ ]:
cmap = sns.diverging_palette(220, 10, as_cmap=True) # one of the many color mappings

# plot the correlation matrix using seaborn
sns.set(style="darkgrid") # one of the many styles to plot using

f, ax = plt.subplots(figsize=(9, 9))

sns.heatmap(df.corr(), cmap=cmap, annot=True)

f.tight_layout()

In [ ]:
%matplotlib inline
from matplotlib import pyplot as plt
sns.set()
sns.pairplot(df, hue="target", size=2)

In [ ]:
sns.set(style="white")

# create a plot grid
g = sns.PairGrid(df[['Age','NoShow','SMSReceived']], diag_sharey=False)
g.map_lower(sns.kdeplot, cmap="Blues_d") # use joint kde on the lower triangle
g.map_upper(plt.scatter) # scatter on the upper
g.map_diag(sns.kdeplot, lw=3) # kde histogram on the diagonal

### Exceptional Work
For part of our effort to provide an example of exceptional work for this project, we decided to create an interactive plot to view the percentage of patients that did not show up for their appointment by each neighborhood location. This plot will exist on a real-scale map of Vitoria - ES, Brazil. This will allow us to make meaningful inferences about distance between clinic neighborhoods since we can view them on a real-life scaled map.

To accomplish this, I leveraged Google Maps Web API and the Bokeh package for creating interactive plots. This requires that a project be registered on Google's API access developer portal using a Google account. Once this was accomplished and an account was properly set up, I received an API key to allow access to Google's Map API from within Python.

Since the data set contained descriptive names for each of the 81 neighborhoods that clinics are located in, I first used Google's API from within Python to search for the location of each neighborhood on Google Maps. This is accomplished by utilizing Google's Geolocation API with the neighborhood name, and specifying that we are only interested in looking for places around Vitoria - ES, Brazil (this is accomplished by appending ', Vitoria - ES, Brazil' to the end of each neighborhood name). The GeoCode data from Google is stashed in a local pickle file for easy retrieval later without the need to access Google's servers.

Two Neighborhoods are excluded from the search; the names are ambiguous (they describe and industrial park and an ocean respectively), and while Google Maps is able to find a location for each of them, the locations it returns are far away from Vitoria Brazil and likely not accurate. Therefore, these two neighborhoods have been excluded from the plot map.

In [12]:
#Note: may need to install library in Anaconda3 dir (C:\Users\<username>\Anaconda3\Scripts\pip.exe install googlemaps)
import pickle
import os
from collections import OrderedDict
import googlemaps
import pandas as pd

geo_pkl = 'data/geo.pkl'

# Access google maps API with API token (created a project with my account)
gmaps = googlemaps.Client(key='AIzaSyCAWI6VE7mlchILCqMYNniGbOMAwAtu_H4')

# Read in raw data
df_raw = pd.read_csv('data/KaggleV2-May-2016.csv')

# 2 ambiguously named neighborhoods will be removed (Google Maps can't find neighborhood location)
df2 = df_raw[df_raw.Neighbourhood != 'PARQUE INDUSTRIAL']
df2 = df2[~df2.Neighbourhood.str.contains('TRINDADE')]

# Get all neighborhood names (sorted alphabetically)
nhoods = sorted(list(df2.Neighbourhood.unique()))

# Get a list of all place names to search for (all are in Vitoria - ES, Brazil)
places = [name + ", Vitoria - ES, Brazil" for name in nhoods]

# Get list of geocodes and save as a pickle object
#  Pickling allows us to limit the number of requests to google API
#  Returns dict of {neighborhood_name: Google_geocode_response (dict)}
print("Getting Geo Data...".ljust(70), end='', flush=True)
if os.path.isfile(geo_pkl):
    with open(geo_pkl, 'rb') as p:
        geocodes = pickle.load(p)
else:
    geocodes = OrderedDict()
    for name in places:
        geocodes[name] = gmaps.geocode(name)
print("Complete!")

# Sometimes API doesn't get a hit (errors out)
#  In this case, it will return list of length 0, so we will re-get any results that are lists with length 0
while 0 in [len(g) for g in geocodes.values() if type(g) == list]:
    for name in geocodes:
        if type(geocodes[name]) == list and len(geocodes[name]) == 0:
            geocodes[name] = gmaps.geocode(name)
            # Check if still missing, and prompt input if it is
            if type(geocodes[name]) == list and len(geocodes[name]) == 0:
                input("%s can't be found. Enter to continue (Ctrl+C to abort)..." % name)

# Some locations have multiple matches in Vitoria - ES, Brazil
#   If that is the case, use only the 1st location
for name in geocodes:
    if type(geocodes[name]) == list:
        geocodes[name] = geocodes[name][0]

# Dump output geocodes to pickle file
with open(geo_pkl, 'wb') as p:
    pickle.dump(geocodes, p)

Getting Geo Data...                                                   Complete!


After importing the data, I used Pandas to calculate the percentage of No-Shows out of the total number of appointments at each of the 79 neighborhoods. Then Bokeh was used to plot color-coded dots on an interactive map representing all clinics in each neighborhood.

Since Google Maps is being used to supply the background of the plot, the Cartesian coordinates on the plot correspond with latitude/longitude coordinates. The previously scraped data from Google Maps provides the coordinates for each location, which can be used as x/y points to plot onto the map.

Bokeh provides the ability to show tooltips with custom information when mousing over points on the plot. This functionality is used to provide the neighborhood name, latitude/longitude coordinates, and the percentage of appointments that are no-shows in that neighborhood when a user mouses over each point on the map.

Finally, Bokeh allows the use of color-scaled palettes to represent the relative magnitudes of custom variables in the plot. I use this functionality to provide a color scale representing no-show percentages, where neighborhood dots become darker orange and red as the no-show percentage becomes higher.

In [13]:
# Plot interactively with Bokeh!
from bokeh.io import output_notebook, show
from bokeh.models import ColumnDataSource, GMapOptions, HoverTool, WheelZoomTool, LogColorMapper
from bokeh.plotting import gmap
from bokeh.palettes import OrRd6 as palette

# Set up the palette that will be used for the color scaling
palette.reverse()
color_mapper = LogColorMapper(palette=palette)

# Output to this notebook for the interactive JS plots
output_notebook()

# Define Google Maps plotting object
map_ops = GMapOptions(lat=-20.285,
                      lng=-40.315,
                      map_type='roadmap',
                      zoom=13
                      )
p = gmap('AIzaSyCAWI6VE7mlchILCqMYNniGbOMAwAtu_H4',
         map_ops,
         title='Vitoria Clinics No-Show Percentage by Neighborhood'
         )

# Get percentages of No-shows for each location
l1 = df2.groupby(['Neighbourhood','No-show']).count()
l2 = df2.groupby(('Neighbourhood')).count()
l3 = l1 / l2 * 100
l3 = l3.iloc[:,:1].rename({'Age': 'PctNoShow'}, axis='columns').reset_index()
no_show_pcts = list(l3[l3['No-show'] == 'Yes']['PctNoShow'])

# Define coordinate data
dat = ColumnDataSource(data=dict(lat=[g['geometry']['location']['lat'] for g in geocodes.values()],
                                 lng=[g['geometry']['location']['lng'] for g in geocodes.values()],
                                 name=[name[:-22].title() for name in geocodes],
                                 noshows=no_show_pcts
                                 )
                       )

# Paint color-scaled circles at respective coordinates
p.circle(x='lng',
         y='lat',
         size=15,
         fill_color={'field': 'noshows',
                     'transform': color_mapper
                     },
         fill_alpha=0.8,
         source=dat
         )

# Add info tooltips to mouse hover
tips=[("Name", "@name"), ("Coords", "(@lng, @lat)"), ("No Shows", "@noshows%")] 
p.add_tools(HoverTool(tooltips=tips))

# Display the plot
show(p)

Loading BokehJS ...

As you can see from the interactive plot above, Bokeh allowed us to create a dynamic, interactive map of the clinic neighborhoods in this data set. I believe this can be very useful for EDA in projects like these, because reading a table of 81 neighborhood names and their corresponding responses doesn't facilitate finding patterns in the data nearly as well as an interactive plot does. The Bokeh plot is fully interactive; it can be zoomed and panned with a mouse to navigate the data in whatever way is the most useful and convenient for the user.

This process required extensive trial and error using Pandas and Bokeh to prepare data for plotting and to create the plot, and  work was also done to make the plot aesthetically pleasing by researching Bokeh's options for presenting the plotted data. We believe that the creation of this plot is an example of exceptional work that has been done to improve this project. This work will be useful to our team, and could also be useful as an example and template for other students wishing to create interactive location plots in the future.